### PR TITLE
Added [KB's]

### LINKED ISSUE #163: [KB] Add KB for prisma-cloud-shiftleft/iac-scan-action
harden-runner-link: 
https://app.stepsecurity.io/github/arjundashrath/StepSecurity_Tests/actions/runs/1716848005

### LINKED ISSUE #262: [KB] Add KB for ChrisCarini/intellij-platform-plugin-verifier-action
Knowledge Base is missing for ChrisCarini/intellij-platform-plugin-verifier-action.

### LINKED ISSUE #263: [KB] Add KB for dependabot/fetch-metadata
harden-runner-link: 
https://app.stepsecurity.io/github/mjpieters/SO-userscripts/actions/runs/1778030458

### LINKED ISSUE #264: [KB] Add KB for dependabot/fetch-metadata
harden-runner-link: 
https://app.stepsecurity.io/github/mjpieters/SO-userscripts/actions/runs/1778030521

### PR DESCRIPTION
**Pull request by Devils-Knight**
Fixes #263 , Fixes #264 
Fixes #262 
Fixes #163

### CODE REVIEW

**Comment by Devils-Knight** [comment]
Issues that already have KB's
Closes #213 
Closes #211 
Closes #173

File: knowledge-base/ChrisCarini/intellij-platform-plugin-verifier-action/action-security.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+name: 'IntelliJ Platform Plugin Verifier'
+github-token:
+  environment-variable-name: GITHUB_TOKEN
+  permissions:
+    contents: read
+    contents-reason: to avoid github rate limiting #Reference: https://github.com/ChrisCarini/intellij-platform-plugin-verifier-action#github-token-authentication
```
Comment (ChrisCarini) [inline knowledge-base/ChrisCarini/intellij-platform-plugin-verifier-action/action-security.yml:6]: As author of this GitHub action, I can confirm this assessment is correct. Thank you @Devils-Knight !

**Review by ChrisCarini — APPROVED** [review]
LGTM!

File: knowledge-base/ChrisCarini/intellij-platform-plugin-verifier-action/action-security.yml (hunk 1 above)
Comment (varunsh-coder) [reply knowledge-base/ChrisCarini/intellij-platform-plugin-verifier-action/action-security.yml:6]: Thanks @ChrisCarini for the comment! We have been thinking how to get authors of GitHub Actions more involved in this project. It gets hard to understand the reason for permissions. And there is always a risk that the permissions may change over time, and this repo may get outdated. Any ideas? How did you find this repo?

Comment (varunsh-coder) [inline knowledge-base/ChrisCarini/intellij-platform-plugin-verifier-action/action-security.yml:6]: Actually if this is for rate limiting only, we should not need `contents: read`. It should only need metadata permissions, which are already there. So, you can remove the permissions part...

File: permissions.go [hunk 2]
```diff
@@ -368,6 +368,33 @@ func (jobState *JobState) getPermissionsForRunStep(step Step) ([]string, error)
 		}
 	}
 
+	// Dependabot auto approve
+	if step.Env["GITHUB_TOKEN"] != "" && isGitHubToken(step.Env["GITHUB_TOKEN"]) {
+		if strings.Contains(runStep, "gh pr review --approve") {
+			permissions = append(permissions, "pull-requests: write")
```
Comment (varunsh-coder) [inline permissions.go:374]: @Devils-Knight can you please move all references of permissions, e.g. `pull-requests: write` to be a constant? That will help avoid typos in the future.

File: permissions.go [hunk 3]
```diff
@@ -368,6 +368,33 @@ func (jobState *JobState) getPermissionsForRunStep(step Step) ([]string, error)
...
+			return permissions, nil
+		}
+	}
+
+	// Dependabot auto merge
```
Comment (varunsh-coder) [inline permissions.go:379]: @Devils-Knight please add a comment with link/ reference to the action, like we do if logic is in the action-security.yml file.

**Comment by Devils-Knight** [comment]
@varunsh-coder I have made the changes requested by you, please review them. And also could you please review the added permissions for [prisma-cloud-shiftleft/iac-scan-action](https://github.com/step-security/secure-workflows/pull/266/files#diff-531d6d3ee9b677e0f250fca1222bc57ea6be5046420ab329e15d1ab36eeeea0c), I am unsure about the permissions to be set for `create_pull_request_check` condition.

File: knowledge-base/ChrisCarini/intellij-platform-plugin-verifier-action/action-security.yml (hunk 1 above)
Comment (ChrisCarini) [reply knowledge-base/ChrisCarini/intellij-platform-plugin-verifier-action/action-security.yml]: > Thanks @ChrisCarini for the comment!

No problem! :)

> We have been thinking how to get authors of GitHub Actions more involved in this project.

Have you considered trying to proactively reach out to GH Action authors and let them know about your project / company and the benefits they would get from onboarding? Perhaps there is a way to get the "top N" GitHub actions used within workflows (of course, excluding ones already accounted for in this repo) and go from there.

> It gets hard to understand the reason for permissions.

Yeah, I can understand that!

> And there is always a risk that the permissions may change over time, and this repo may get outdated.

Ironic, as this was actually my first thought after I left the comment/approval. "What if I was a bad-actor, secretly, and I got my GitHub action added to this repo, but then decided to change the intent behind the usage of the GH token env-var? That wouldn't be good. How would the 'good-actors' (ie, step-security, in this case) know or get that information propagated."

> Any ideas?

Perhaps excessive, but would it be possible for your application to keep track of the GH actions it is aware of, and monitor for any changes that might add/remove the usage of the `GITHUB_TOKEN` (or other sensitive secret) and then open an issue with the respective information to re-evaluate accordingly?

> How did you find this repo?

I was guided to https://app.stepsecurity.io/ after recently onboarding to OSSF's https://github.com/ossf/scorecard-action GitHub action. From there, I pasted in some of my workflows on my repos (of which some of them use a GitHub action I wrote / maintain myself), and saw that nothing was changed, but the web app told me a ticket was created. Clicking that link lead me to the GH issue, and thus this PR.

Comment (varunsh-coder) [reply knowledge-base/ChrisCarini/intellij-platform-plugin-verifier-action/action-security.yml]: Thanks @ChrisCarini this is very valuable feedback. Let me make a plan on proactively reaching out to GitHub Action authors.

File: permissions.go [hunk 4]
```diff
@@ -362,12 +392,40 @@ func (jobState *JobState) getPermissionsForRunStep(step Step) ([]string, error)
...
-				permissions = append(permissions, "packages: write")
+				permissions = append(permissions, packages_write)
 				return permissions, nil
 			}
 		}
 	}
 
+	// Dependabot run steps reference: https://github.com/dependabot/fetch-metadata#usage-instructions
+	// Dependabot auto approve
+	if step.Env["GITHUB_TOKEN"] != "" && isGitHubToken(step.Env["GITHUB_TOKEN"]) {
```
Comment (varunsh-coder) [inline permissions.go:403]: @Devils-Knight please add test cases for these, else the test coverage will reduce. You can just create relevant files in the input and output folders so these lines get covered.

**Comment by varunsh-coder** [comment]
> @varunsh-coder I have made the changes requested by you, please review them. And also could you please review the added permissions for [prisma-cloud-shiftleft/iac-scan-action](https://github.com/step-security/secure-workflows/pull/266/files#diff-531d6d3ee9b677e0f250fca1222bc57ea6be5046420ab329e15d1ab36eeeea0c), I am unsure about the permissions to be set for `create_pull_request_check` condition.

LGTM

File: testfiles/joblevelpermskb/output/dependabot-run-steps.yml [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+          github-token: "${{ secrets.GITHUB_TOKEN }}"
+      - name: Approve a PR
+        run: gh pr review --approve "$PR_URL"
+        env:
+          PR_URL: ${{github.event.pull_request.html_url}}
+          GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}
+
+  Dependabot auto-merge:
+    permissions:
+      contents: write
```
Comment (varunsh-coder) [inline testfiles/joblevelpermskb/output/dependabot-run-steps.yml:37]: @Devils-Knight I realized we are not adding reasons when permissions get added via code. Will you be able to fix this?

Comment (Devils-Knight) [reply testfiles/joblevelpermskb/output/dependabot-run-steps.yml]: @varunsh-coder I have fixed the issue, Please review the changes.

**Pull request merged**